interop-testing/src/test/java/io/grpc/testing/integration/ProxyTest.java

```diff
@@ -0,0 +1,230 @@
+package io.grpc.testing.integration;
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketImpl;
+import java.net.UnknownHostException;
+
+@RunWith(JUnit4.class)
+public class ProxyTest {
+
+  private int serverPort = 5001;
+  private int proxyPort = 5050;
+
+  // These all work individually right now but there is running them sequentially causes errors
+  // because shut down is not clean yet.
+
+  @Test
+  @Ignore
+  public void smallLatency() throws UnknownHostException, IOException, InterruptedException {
+    Server server = new Server();
+    (new Thread(server)).start();
+
+    int latency = 10;
+    TrafficControlProxy p = new TrafficControlProxy();
+    p.setLatency(latency);
+    (new Thread(new Proxy(p))).start();
+    Thread.sleep(100);
+    Socket client = new Socket("127.127.127.127", proxyPort);
+    DataOutputStream clientOut = new DataOutputStream(client.getOutputStream());
+    DataInputStream clientIn = new DataInputStream(client.getInputStream());
+    byte[] message = new byte[1];
+
+    // warm up
+    for (int i = 0; i < 10; i++) {
+      clientOut.write(message, 0, 1);
+      clientIn.read(message);
+    }
+    // test
+    long start = System.currentTimeMillis();
+    clientOut.write(message, 0, 1);
+    clientIn.read(message);
+    long stop = System.currentTimeMillis();
+
+    long rtt = (stop - start);
+    // NOTE: arbitrarily picked acceptable error to be 2 milliseconds
+    assertEquals(latency, rtt, 2);
+
+    server.shutDown();
+    p.shutDown();
+    client.close();
+  }
+
+  @Test
+  @Ignore
+  public void bigLatency() throws UnknownHostException, IOException, InterruptedException {
+    Server server = new Server();
+    (new Thread(server)).start();
+
+    int latency = 500;
+    TrafficControlProxy p = new TrafficControlProxy();
+    p.setLatency(latency);
+    (new Thread(new Proxy(p))).start();
+    Thread.sleep(100);
+    Socket client = new Socket("127.127.127.127", proxyPort);
+    DataOutputStream clientOut = new DataOutputStream(client.getOutputStream());
+    DataInputStream clientIn = new DataInputStream(client.getInputStream());
+    byte[] message = new byte[1];
+
+    // warm up
+    for (int i = 0; i < 10; i++) {
+      clientOut.write(message, 0, 1);
+      clientIn.read(message);
+    }
+    // test
+    long start = System.currentTimeMillis();
```

Comment:
don't use currentTimeMillis.  It can do funky things like go backwards in time.  Use nanoTime unless you know that cTM is really what you want.